torch/fx/experimental/proxy_tensor.py

```diff
@@ -70,12 +70,14 @@ def maybe_disable_fake_tensor_mode():
         return nullcontext()
 
 
-def proxy_call(func_overload, args, kwargs=None):
+def proxy_call(func_overload, dispatch_mode, args, kwargs=None):
     if kwargs is None:
         kwargs = {}
     func = func_overload.overloadpacket
     if func_overload in CURRENT_DECOMPOSITION_TABLE:
-        return CURRENT_DECOMPOSITION_TABLE[func_overload](*args, **kwargs)
+        proxy_mode = enable_torch_dispatch_mode(dispatch_mode) if dispatch_mode else nullcontext()
+        with proxy_mode, torch.overrides.enable_reentrant_dispatch():
```

Comment:
Now I'm a little puzzled... how come we need `enable_reentrant_dispatch`? Based on what you've told us, it seems to be required, but I'm blanking on exactly why mechanically it's needed. In particular, if there is a dispatch mode stack, we don't seem to require people to reenable reentrant dispatch when they call into func (which will ostensibly go to the next dispatch mode in the stack).

Reply:
This doesn't smell right; either there's another bug lurking, or we need to turn these two context managers into a consolidated context manager to make it easier for people to do this correctly.

Reply:
I think we want to use `enable_reentrant_dispatch` so that we can reset the higher dispatch bits, in order to trace through CompositeImplicitAutograd ops. (On the first dispatch, PythonTLSSnapshot saves the original DispatchKeySet so that `enable_reentrant_dispatch` can restore the same DispatchKeySet).

re: turning these two context managers into a consolidated context manager - we could make `enable_reentrant_dispatch` take an optional torch_dispatch object? But there are also cases like this: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/composite_compliance.py#L200-L205 where `enable_reentrant_dispatch()` is sometimes getting called via `enable_torch_dispatch_mode()` on a tensor subclass, where it might be difficult to figure out which torch_dispatch mode to use. (see failures on #81677)

Reply:
OK I take it back, what's going on here is probably wrong and we should find some other mechanism for this.

If you are using `enable_reentrant_dispatch` to shunt as back to CompositeImplicitAutograd, not only are we reenabling the composites, but we're also reenabling autograd, which can lead to weird things where we start building up grad fn, even though at the outer layer we've already set autograd. Let me patch your change in and see if I can figure out how to do it better.

Reply:
OK, this is probably the easiest way to solve this for now.

Long term what would make more sense is for proxy tensor to be able to directly make use of CompositeImplicitAutograd registered kernels *in its body itself*. The same way that when a CPU kernel calls a composite kernel, that will hit the composite registration in CPU itself (without reenabling reentrant dispatch) and eventually getting to the real CPU backend kernels. Also the same way we consult a decomposition table here. We just have no convenient way to actually get at these registrations at the moment; @anjali411's proposal to get out ops from the dispatch table would help a little, but we would need to be able to query for composites specifically.

Changing https://github.com/pytorch/pytorch/issues/81682 would also help, as the tracing kernel would then be a fallback kernel and defer to composites.

Reply:
Another way of fixing this, by @albanD, is add a composite key separate from autograd which doesn't get disabled when we go in. cc @zou3519